draco/synthesis/expand_perturbed.py

```diff
@@ -0,0 +1,598 @@
+
+        # Define a BeamPerturbation container to hold the beam perturbations
+        perturbation_list = containers.BeamPerturbation(freq = freqmap, input = pertlistlen,
+                            distributed = True, prod = desired_prod, comm = map_.comm)
+        # Add beam perturbation values to container.
+        perturbation_list.pert[:] = perturbations
+
+        return perturbation_list
+
+
+class GenerateSinglePerturbation(task.SingleTask):
+    """ Generate small, random number perturbation for just 1 input to send to the
+        ExpandPerturbedProducts task.
+
+        This is a bit limited in scope, but useful for getting an idea of where
+        the perturbation is showing up in later analyses.
+    """
+
+    # Define multiplier value. This is usually set in the .yaml file used to run the pipeline.
+    # The idea is it sets the overall size of the beam perturbations, allowing
+    # us to ensure smaller values than the usual size of a numpy random number.
+    pert_val = config.Property(proptype = float, default = 0.01)
+    pert_index = config.Property(proptype = float, default = 0)
+    def setup(self,telescope):
+        """Get a reference to the telescope class.
+
+        Parameters
+        -------------
+        tel : : class: `drift.core.TransitTelescope`
+            Telescope object.
+        """
+        self.telescope = telescope
+
+    def process(self, sstream, map_):
+        # Need to write container and add saving to container to this.
+        # Then will be ready to roll out in first form.
+        # Drop multiplier if that's not working
+        """ Generate perturbations for each feed.
+        """
+        tel = self.telescope
+
+        # Determine total number of inputs from sstream
+        ninput = len(sstream.input)
+        # Determine the number of inputs per perturbation.
+        ninputperpert = ninput / tel.npert
+
+        # There's a slight chance this is going to freak out for more than 1
+        # perturbation. If it does, we'll modify it.
+        pertlistlen = ninputperpert
+        #print "Pert List Len", pertlistlen
+
+        # Define frequency map for upcoming container.
+        freqmap = map_.index_map['freq'][:]
+
+        # Calculate random numbers to be perturbation values. Set the general size using input mult.
+        #perturbations = np.random.standard_normal(pertlistlen) * self.mult
+        # Do this only for communicator rank 0 (the first one) so that you don't
+        # have four independent sets of perturbations running around.
+        comm = sstream.comm
+        if comm.rank == 0:
+            # Set the perturbation matrix to be all zeros
+            perturbations = np.zeros(pertlistlen)
+            # Set just one entry (which you selected) to be the perturbation value you put in.
+            perturbations[self.pert_index] = self.pert_val
+        else:
+            perturbations = None
+
+        perturbations = comm.bcast(perturbations, root=0)
```

Comment:
I don't think I fully understand the code above. It seems to me you define an array of zeros then change only the first value of it. Than broadcast this array to all communicators. What is the use of the rest of the array 'perturbation' when 'pertlistlen' is not 1?